tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/auto_query_fixer/QueryTokenService.java

```diff
@@ -0,0 +1,85 @@
+package com.google.cloud.bigquery.utils.auto_query_fixer;
+
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.parser.babel.SqlBabelParserImpl;
+import org.apache.calcite.sql.parser.babel.Token;
+import org.apache.calcite.util.SourceStringReader;
+import org.apache.commons.lang3.tuple.Pair;
+
+import com.google.cloud.bigquery.utils.auto_query_fixer.entity.IToken;
+import com.google.cloud.bigquery.utils.auto_query_fixer.entity.TokenImpl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryTokenService {
+
+  public Pair<IToken, IToken> getNearbyTokens(String sql, int beginLine, int beginColumn) {
+    IToken previous = null;
+
+    for (IToken token : getAllTokens(sql)) {
+      if (token.getBeginLine() >= beginLine && token.getBeginCol() >= beginColumn) {
+        return Pair.of(previous, token);
+      }
+      previous = token;
+    }
+
+    return Pair.of(previous, previous);
+  }
+
+  public List<IToken> getAllTokens(String sql) {
+    List<IToken> tokens = new ArrayList<>();
+
+    // a parser contains token manager, which will be used to tokenize sql.
+    final SqlParser.ConfigBuilder configBuilder =
+        SqlParser.configBuilder()
+            .setParserFactory(SqlBabelParserImpl.FACTORY);
+    SqlBabelParserImpl parserImpl = (SqlBabelParserImpl) configBuilder.build().parserFactory().getParser(new SourceStringReader(sql));
+
+    Token token;
+    while ((token = parserImpl.getNextToken()).kind != 0) {
```

Comment:
Curious what is 'kind' here? why we can use this as termination condition?

Reply:
modify to

https://github.com/GoogleCloudPlatform/bigquery-utils/pull/73/commits/0f80bd6fc66697b00f278c0cc5ecee9063fb7639#diff-f420e24e3bb7f22da366200956de476aR57